### ISSUE TITLE
Many reduction operators do not support reducing over multiple dimensions

### ISSUE THREAD

**Issue by heitorschueroff** [pytorch/pytorch#61582]
The following reduction operators should be updated to allow reducing over multiple dimensions. Note that adding support for reducing over multiple dimensions to `torch.{min,max,median,mode} is blocked on https://github.com/pytorch/pytorch/issues/61490. And operators that currently accept both `None` and int for the parameter `dim` in the same signature are blocked by https://github.com/pytorch/pytorch/issues/44409.

- [ ] [torch.all](https://pytorch.org/docs/master/generated/torch.all.html#torch.all)
- [ ] [torch.any](https://pytorch.org/docs/master/generated/torch.any.html#torch.any)
- [ ] [torch.max](https://pytorch.org/docs/master/generated/torch.max.html#torch.max)
- [ ] [torch.min](https://pytorch.org/docs/master/generated/torch.min.html#torch.min)
- [ ] [torch.median](https://pytorch.org/docs/master/generated/torch.median.html#torch.median)
- [ ] [torch.nanmedian](https://pytorch.org/docs/master/generated/torch.nanmedian.html#torch.nanmedian)
- [ ] [torch.mode](https://pytorch.org/docs/master/generated/torch.mode.html#torch.mode)
- [ ] [torch.prod](https://pytorch.org/docs/master/generated/torch.prod.html#torch.prod)
- [ ] [torch.quantile](https://pytorch.org/docs/master/generated/torch.quantile.html#torch.quantile)
- [ ] [torch.nanquantile](https://pytorch.org/docs/master/generated/torch.nanquantile.html#torch.nanquantile)

**Related Issues**
- [ ] https://github.com/pytorch/pytorch/issues/56586
- [ ] https://github.com/pytorch/pytorch/issues/53900
- [ ] https://github.com/pytorch/pytorch/issues/61453

**Comment by OliverFM** [comment]
Hey!
@ashlilies and I were interested in working on this issue. We are both new to contributing to torch though -- and weren't able to track down the definition of `torch.all` could anyone point us in the right direction?

**Comment by vadimkantorov** [comment]
Just stumbled on the case with torch.any - still hurts :)

**Comment by vadimkantorov** [comment]
@OliverFM A recent PR touching torch.all may help out: https://github.com/pytorch/pytorch/pull/64642/files

**Comment by vadimkantorov** [comment]
It also seems that underlying `allany_impl` may already support multidim, but maybe some bindings are missing ...

**Comment by vadimkantorov** [comment]
Stumbled on quantile not supporting multidim :(

**Comment by thomasahle** [comment]
Stumbled on median not supporting multidim. Is there an alternative way to do batch medians without a for loop?

**Comment by ablacan** [comment]
> Stumbled on quantile not supporting multidim :(

Did you ever find a way to bypass this ?

**Comment by vadimkantorov** [comment]
I did manual flatten of the interesting dimensions, but it's not very elegant

**Comment by alisterburt** [comment]
@vadimkantorov this flattening also causes a memory allocation which explodes memory requirements in certain cases. Below is an example with a median calculation over a 3D sliding window view of a `(d, h, w)` shaped tensor


```python
import torch
import einops

def sliding_window_3d(tensor, window_size):
    d, h, w = tensor.shape
    window_depth, window_height, window_width = window_size

    # Calculate the number of windows in each dimension
    new_d = d - window_depth + 1
    new_h = h - window_height + 1
    new_w = w - window_width + 1

    # Calculate the new strides
    original_strides = tensor.stride()
    new_strides = (original_strides[0], original_strides[1], original_strides[2],
                   original_strides[0], original_strides[1], original_strides[2])

    # Create the sliding window view
    window_view = torch.as_strided(tensor, (new_d, new_h, new_w, window_depth, window_height, window_width),
                                   new_strides)
    return window_view


# Example usage
tensor = torch.arange(32 * 32 * 32).reshape(32, 32, 32)
window_size = (3, 3, 3)
window_view = sliding_window_3d(tensor, window_size)
print(window_view.shape)  # (30, 30, 30, 3, 3, 3)

# make the window 1d for median calc
window_view_1d = einops.rearrange(window_view, 'd h w wd wh ww -> d h w (wd wh ww)')  # reallocates here

# calculate the median
median = torch.median(window_view_1d, dim=-1)
print(median.values.shape)  # 30, 30, 30
```

**Comment by vadimkantorov** [comment]
IIUC, this should not work properly near the boundaries, right? maybe this can be worked around by pad + unfold/stride-tricks. a related issue: https://github.com/pytorch/pytorch/issues/44991

**Comment by alisterburt** [comment]
@vadimkantorov right, that implementation is naive to boundaries and padding would alleviate that. 

The problem is that reshaping the view results in the allocation of an ~n*d tensor where n are the number of elements of the original tensor and d are the number of elements in the window - if the window is large on a 3D input this quickly becomes limiting

**Comment by vadimkantorov** [comment]
I also wonder if any classical image processing fast median filter impls are relevant here (might be useful if strides are small and windows overlap a lot)

**Comment by jscanvic** [comment]
It'd be very handy if `torch.aminmax` could operate on multiple dimensions like `torch.amin` and `torch.amax`.
